ipcc/src/handle.rs

```diff
@@ -0,0 +1,107 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2023 Oxide Computer Company
+
+use std::{
+    ffi::{c_int, CStr, CString},
+    ptr,
+};
+
+use crate::IpccError;
+use crate::{ffi::*, IpccErrorInner};
+
+pub struct IpccHandle(*mut libipcc_handle_t);
+
+impl Drop for IpccHandle {
+    fn drop(&mut self) {
+        unsafe {
+            libipcc_fini(self.0);
+        }
+    }
+}
+fn ipcc_fatal_error<C: Into<String>>(
+    context: C,
+    lerr: libipcc_err_t,
+    syserr: c_int,
+    errmsg: CString,
+) -> IpccError {
+    let context = context.into();
+    let syserr = if syserr == 0 {
+        "no system errno".to_string()
+    } else {
+        std::io::Error::from_raw_os_error(syserr).to_string()
+    };
+    let inner = IpccErrorInner {
+        context,
+        errmsg: errmsg.to_string_lossy().into_owned(),
+        syserr,
+    };
+    match lerr {
+        LIBIPCC_ERR_OK => panic!("called fatal on LIBIPCC_ERR_OK"),
+        LIBIPCC_ERR_NO_MEM => IpccError::NoMem(inner),
+        LIBIPCC_ERR_INVALID_PARAM => IpccError::InvalidParam(inner),
+        LIBIPCC_ERR_INTERNAL => IpccError::Internal(inner),
+        LIBIPCC_ERR_KEY_UNKNOWN => IpccError::Internal(inner),
```

Comment:
should this be `IpccError::KeyUnknown`?

Reply:
Nice catch -- dang it copy and paste!
Fixed in 77bd065.